alg/sha256_arm.c

```diff
@@ -87,64 +58,45 @@ static const uint32_t Krnd[64] = {
  */
 #ifdef POSIXFAIL_ABSTRACT_DECLARATOR
 void
-SHA256_Transform_arm(uint32_t state[8], const uint8_t block[64],
-    uint32_t W[64], uint32_t S[8])
+SHA256_Transform_arm(uint32_t state[8], const uint8_t block[64])
 #else
 void
 SHA256_Transform_arm(uint32_t state[static restrict 8],
-    const uint8_t block[static restrict 64], uint32_t W[static restrict 64],
-    uint32_t S[static restrict 8])
+    const uint8_t block[static restrict 64])
 #endif
 {
+	uint32x4_t Y[4];
+	uint32x4_t S[2]; /* S[0] and S[1] alternate between ABEF and CDGH. */
```

Comment:
I'm confused by the comment here.  It looks like we're loading ABCD into one and EFGH into the other; is there some permutation I'm missing?

Reply:
I'm confused too, but I'm now fairly confident that it's because I misunderstood the "rotating state" in our generic SHA256 code when I was first working on this.

Looking at our first test case, the SHA256 of "", with our generic `sha256.c` (after removing the "if hwaccel, do this other function") code gives:
```
after 4 rounds
3037e1b5 6fdb22fa 2001b256 18c7e2a2 4c23eacb c4ac5526 9ce564d0 7c08884d 
after 8 rounds
7ad5b1b6 f0205590 292e3b42 4366f730 53f94dcb 82ede06f 4f6540fa e7d0641e
```
That's from printing `S[j]` from 0 to 7.

The Arm code, with slightly different formatting (sorry, could change if you want), gives:
```
after 4 rounds
S[0] 4c23eacb c4ac5526 9ce564d0 7c08884d
S[1] 3037e1b5 6fdb22fa 2001b256 18c7e2a2
after 8 rounds
S[0] 7ad5b1b6 f0205590 292e3b42 4366f730
S[1] 53f94dcb 82ede06f 4f6540fa e7d0641e
```

In other words, if we interpret the generic S[8] as ABCDEFGH, and the Arm S[2] as ABCD, EFGH; then the two implementations agree after 8 rounds.  But after 4 rounds, the Arm implementation is flipped, with S[0] containing what we'd expect to see in S[1] and vice versa.

**However**, future reading about the Arm instructions, and looking at our generic code, makes me think that I got the "alternating" thing the wrong way around.  Namely, it looks like the Arm code always stores ABCD in S[0]; but the generic code rotates the state.  So after 4 iterations, the generic code has S[0] = EFGH; after 8, it's back to S[0] = ABCD.

In my defense, our SHANI implemention does have the state flipping between S[0] and S[1].  Namely:
```
        /* Perform two rounds of SHA256, using the low two words in M. */       \
        S[1] = _mm_sha256rnds2_epu32(S[1], S[0], M);                            \
                                                                                \
        /* Shift the two words of M down and perform the next two rounds. */    \
        M = _mm_srli_si128(M, 8);                                               \
        S[0] = _mm_sha256rnds2_epu32(S[0], S[1], M);      
```

In summary, I think that the solution is to change those comments in `sha256_arm.c` to clarify that `S` doesn't have rotating state in the C code; it's all taken care of by the intrinsics.